README.md
# Packet-journey, linux router based on DPDK

The purpose of this project is to provide a **free** application capable of:

* Switching many packets using the LPM algorithm
* Making this switching scalable with the possibility of adding more packet queues/CPUs
* Learning routes from the Linux kernel using Netlink
* Learning neighbors from Netlink, the kernel is refreshing them automatically
* Being able to forward some packets to the kernel which will handle them (ARP, ICMP, BGP)
* Rate limiting ICMP packets that are sent to the kernel
* Supporting L3/L4 ACLs
* Gathering statistics and managing ACLs through a cli based on a unixsock
* Being configurable through a configfile

Most of these features are based on various DPDK libs.

## Internals

### Threads

Packet-journey (pktj) is composed of multiple kind of threads:

* pktj, the main thread which is doing all the initialization and are handling signals.
* forward-LCOREID, those threads are doing the forwarding part. They are reading packets from the PMD, doing the processing of those packets and sending them back to the PMD.
* kni-LCOREID, those threads are reading packets from the KNI and sending them to the configured port.
* lcore-slave-LCOREID, those threads are doing nothing, they are just waiting.
* control-SOCKETID, those threads are in charge of receiving NETLINK messages from the IFADDR, ROUTE, NEIGH and LINK groups and handling them.
* cmdline-SOCKETID, those threads are presenting a CLI through the unixsocks.
* rdtsc-SOCKETID, those threads are reading the TSC value and exposing it to the lcore-slave threads.

For optimal performances, the forwarding threads must be alone on their cores. All other threads can be scheduled on the same lcore.

### Processing steps

The forwarding threads are running the main_loop() function. It basically follows these steps:

1. Read up to 32 packet descriptors
2. If none, read again
3. Prepare the ACL processing for the new packets and filter them
4. Find the correct neighbor for the remaining packets by looking into the IPv4 or IPv6 LPM
5. If a packet has no possible next_hop in the LPM or if a packet has the router IP, send it to the KNI and remove it from the rest of the processing loop
6. For each of the remaining packets, set the correct destination MAC address according to the selected next_hop
7. Reorder packets by destination port
8. Send the packets in batch grouped by destination port

## Configuration examples

```
pktj -l 0,1,2,3 -n 4 --socket-mem=4096 --log-level=4 -- \
--configfile /root/devel/router-dpdk/tests/integration/lab00/pktj.conf
```

with `pktj.conf` containing:

```
; pktj
[pktj]
callback-setup  = /root/devel/router-dpdk/tests/integration/lab00/up.sh
rule_ipv4       = /root/devel/router-dpdk/tests/integration/lab00/acl.txt
rule_ipv6       = /root/devel/router-dpdk/tests/integration/lab00/acl6.txt
promiscuous     = 1
kni_rate_limit  = 1000
aclavx2         = 1

; Port configuration
[port 0]
eal queues      = 0,1 1,2 ; queue,lcore
kni             = 3,0 ; lcore_tx,kthread
```

These settings will launch the following:

* pktj with 2 forwarding threads, on core 1 and 2,
* a KNI tx thread on core 3, and
* the script ```up.sh``` after setting up the KNI.

The script `up.sh` configures the KNI IPv4, IPv6 and MAC addresses. It also
starts some processes (bgpd and zebra in our case).


```
#!/bin/sh

link1=$1
mac1=$2

ip link set $link1 up
ip link add link $link1 name $link1.2000 type vlan id 2000
ip link set $link1.2000 address $mac1
ip link set $link1.2000 up

ip addr add 1.2.3.5/24 dev $link1.2000
ip route add 1.2.4.0/24 via 1.2.3.4
ip addr add 2001:3::5/48 dev $link1.2000
ip route add 2001:4::/48 via 2001:3::4

# delay start because of quagga racy start
sleep 15s

/usr/lib/quagga/zebra &
/usr/lib/quagga/bgpd &
```

You can try this out using our [QEMU test lab](tests/integration).

## Performances

The test scenario is done between 2 physical machines having both a XL710 NIC
and a E5-2630L CPU. The first host is running Packet-Journey; it receives
traffic from the second host and sends it back.

For simulating real life conditions, the tests are done with 500k
routes (something similar to a BGP full view) and 50 ACL (we use scalar
implementation for ACL since the CPU don't have AVX2 instructions). We have
also a percentage of the packets which are forwarded to the KNI instead of
going through the fastpath.

The packets are generated using DPDK-PktGen which is configured for sending 64
Bytes UDP packets with a random source and destination IP. When
Configured with 4 RX queues, Packet-Journey is able to forward 21773.47 mbits.

The graphs of the test scenario are in the [doc](doc/pktj_test_acl_dpdk_4q).
The easiest graphs to read are the ones from pktgen since it shows you the TX
RX rates: [PktGen report](doc/pktj_test_acl_dpdk_4q/results/0/pktgen-perf_report.html).



debian/changelog
pktj (0.1.34) unstable; urgency=medium

  * Fix vPMDs with jumbo frame enabled

 -- Fatih Acar <fatih@gandi.net>  Mon, 08 Oct 2018 15:19:17 +0200

pktj (0.1.33) unstable; urgency=medium

  * Support for DPDK 18.02
  * Allow jumbo frames through KNI
  * Fix naming issue of dump commands

 -- Fatih Acar <fatih@gandi.net>  Thu, 14 Jun 2018 10:52:52 +0200

pktj (0.1.32) unstable; urgency=medium

  * main: fix endianness issue with ipv4 packets

 -- Fatih Acar <fatih@gandi.net>  Wed, 20 Dec 2017 14:09:29 +0100

pktj (0.1.31) unstable; urgency=medium

  * control: handle default routes

 -- Fatih Acar <fatih@gandi.net>  Wed, 20 Sep 2017 12:48:58 +0200

pktj (0.1.30) unstable; urgency=medium

  * Authorize NUD_STALE update for gratuitous arp.

 -- Nikita Kozlov <nikita@gandi.net>  Tue, 29 Nov 2016 14:20:51 +0100

pktj (0.1.29) unstable; urgency=medium

  * Fix rte_lpm_delete speed.

 -- Nikita Kozlov <nikita@gandi.net>  Sun, 30 Oct 2016 17:41:51 +0200

pktj (0.1.28) unstable; urgency=medium

  * Fix NUD status handling in neighbor4/neighbor6.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 01 Jul 2016 16:26:51 +0200

pktj (0.1.27) unstable; urgency=medium

  * Improve netlink handling and lpm6 speed.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 29 Jun 2016 11:21:51 +0200

pktj (0.1.26) unstable; urgency=medium

  * Use dpdk ubuntu dpdk 16.04 packages.

 -- Nikita Kozlov <nikita@gandi.net>  Sun, 19 Jun 2016 17:46:51 +0200

pktj (0.1.25) unstable; urgency=medium

  * Rewrite control thread spawning logic.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 03 Feb 2016 14:39:51 +0100

pktj (0.1.24) unstable; urgency=low

  * Modify the packaging for building with ubuntu dpdk packages.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 06 Jan 2016 13:41:29 +0100

pktj (0.1.23) unstable; urgency=low

  * Set the correct prefix length for the local route.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 27 Nov 2015 16:12:29 +0200

pktj (0.1.22) unstable; urgency=low

  * Fix packaging of new scripts.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 27 Nov 2015 13:07:29 +0200

pktj (0.1.21) unstable; urgency=low

  * Add BLACKHOLE route type.
  * Fix some coverity warning.
  * Add an SNMP export script.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 27 Nov 2015 11:28:29 +0200

pktj (0.1.20) unstable; urgency=low

  * Decrease usleep time in rdtsc thread.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 28 Oct 2015 13:15:29 +0200

pktj (0.1.19) unstable; urgency=low

  * Increase max number of LPM entry.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 23 Oct 2015 14:52:29 +0200

pktj (0.1.18) unstable; urgency=low

  * Fix ratel and kni_tx stats.
  * Fix a potential sigpipe in cmdline.

 -- Nikita Kozlov <nikita@gandi.net>  Tue, 20 Oct 2015 17:24:29 +0200

pktj (0.1.17) unstable; urgency=low

  * catch SIGTERM and SIGQUIT for terminating properly the processes.
  * add local route for neighbor in the same network as ourselves.
  * licensing.
  * rename rdpdk to packet-journey.
  * add a usleep of 1ms in kni_main_loop for not consuming all the core.
  * fix acces to the kni from multiple threads.
  * make acl switching threadsafe.
  * documentations.

 -- Nikita Kozlov <nikita@gandi.net>  Mon, 19 Oct 2015 16:34:29 +0200

pktj (0.1.16) unstable; urgency=low

  * Fix kni config from file.

 -- Nikita Kozlov <nikita@gandi.net>  Tue, 29 Sep 2015 13:19:29 +0200

pktj (0.1.15) unstable; urgency=low

  * check NUD states in netlink callbacks and add an ipv6 multicast link-local route.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 23 Sep 2015 11:33:29 +0200

pktj (0.1.14) unstable; urgency=low

  * fix ipv6 hop_limit bug and add rate-limiting to the kni.

 -- Nikita Kozlov <nikita@gandi.net>  Tue, 22 Sep 2015 01:08:29 +0200

pktj (0.1.13) unstable; urgency=low

  * fix a bug is acl and ttl decrement.

 -- Nikita Kozlov <nikita@gandi.net>  Mon, 14 Sep 2015 16:08:29 +0200

pktj (0.1.12) unstable; urgency=low

  * fix ipv6 local route bug.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 09 Sep 2015 00:37:29 +0200

pktj (0.1.11) unstable; urgency=low

  * Use prctl for tracking childs + cmdline polling.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 02 Sep 2015 17:40:29 +0200

pktj (0.1.10) unstable; urgency=low

  * fix bug in kni multithreading.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 28 Aug 2015 14:27:29 +0200

pktj (0.1.9) unstable; urgency=low

  * fix jumbo params.

 -- Nikita Kozlov <nikita@gandi.net>  Thu, 27 Aug 2015 04:11:29 +0200

pktj (0.1.8) unstable; urgency=low

  * Add missing libs in package.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 26 Aug 2015 13:59:29 +0200

pktj (0.1.7) unstable; urgency=low

  * Dynamicly link against newest dpdk.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 26 Aug 2015 12:59:29 +0200

pktj (0.1.6) unstable; urgency=low

  * Make it compatible with dpdk 2.1.0 .

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 26 Aug 2015 00:42:29 +0200

pktj (0.1.5) unstable; urgency=low

  * fix rename vEth to dpdk.

 -- Nikita Kozlov <nikita@gandi.net>  Tue, 25 Aug 2015 02:23:29 +0200

pktj (0.1.4) unstable; urgency=low

  * rename vEth to dpdk.

 -- Nikita Kozlov <nikita@gandi.net>  Thu, 20 Aug 2015 14:23:29 +0200

pktj (0.1.3) unstable; urgency=low

  * fix typo in rpath for sandybridge and ivybridge.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 19 Aug 2015 23:37:29 +0200

pktj (0.1.2) unstable; urgency=low

  * add back vlan stripping on rx.

 -- Nikita Kozlov <nikita@gandi.net>  Wed, 10 Aug 2015 03:03:29 +0200

pktj (0.1.1) unstable; urgency=low

  * add DEVICE_BOUND_OPT and DEVICE_UNBOUND_OPT in init script.

 -- Nikita Kozlov <nikita@gandi.net>  Thu, 06 Aug 2015 15:39:29 +0200

pktj (0.1.0) unstable; urgency=low

  * Initial release.

 -- Nikita Kozlov <nikita@gandi.net>  Fri, 31 Jul 2015 15:59:16 +0100


app/config.h
/*
 * Packet-journey userland router which uses DPDK for its fastpath switching
 *
 */
/*
 * Copyright (c) 2015 Gandi S.A.S.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#ifndef __PKTJ_CONFIG_H
#define __PKTJ_CONFIG_H

void print_usage(const char* prgname);
int parse_args(int argc, char** argv);
uint8_t get_port_n_rx_queues(uint8_t port);
void rate_limit_config_from_file(const char* file_name);
int rate_limit_address(cmdline_ipaddr_t* ip, uint32_t num, int socket_id);

extern uint16_t nb_lcore_params;
extern struct lcore_params* lcore_params;
extern uint32_t enabled_port_mask;
extern int promiscuous_on;
extern int numa_on;
extern uint32_t kni_rate_limit;
extern const char* callback_setup;
extern const char* unixsock_path;
extern const char* ratelimit_file;
extern struct rte_eth_conf port_conf;

struct lcore_params {
	uint8_t port_id;
	uint8_t queue_id;
	uint8_t lcore_id;
} __rte_cache_aligned;

#define MAX_RX_QUEUE_PER_LCORE 16
#define MAX_TX_QUEUE_PER_PORT RTE_MAX_ETHPORTS
#define MAX_RX_QUEUE_PER_PORT 128

struct lcore_rx_queue {
	uint8_t port_id;
	uint8_t queue_id;
} __rte_cache_aligned;

struct mbuf_table {
	uint16_t len;
	struct rte_mbuf* m_table[MAX_PKT_BURST];
};

#define INVALID_RLIMIT_RANGE UINT8_MAX
#define MAX_RLIMIT_RANGE 10  // max number of cidr ranges that can be limited

#define RLIMIT_RANGE 16  // group addresses by /16
#define MAX_RLIMIT_RANGE_NET \
	(1 << RLIMIT_RANGE)  // max number of subnets for that prefix
#define MAX_RLIMIT_RANGE_HOST \
	(1 << (32 - RLIMIT_RANGE))  // max number of hosts in a range

struct lcore_conf {
	uint16_t n_rx_queue;
	struct lcore_rx_queue rx_queue_list[MAX_RX_QUEUE_PER_LCORE];
	uint16_t tx_queue_id[RTE_MAX_ETHPORTS];
	struct mbuf_table tx_mbufs[RTE_MAX_ETHPORTS];
	lookup_struct_t* ipv4_lookup_struct;
	lookup6_struct_t* ipv6_lookup_struct;
	neighbor_struct_t* neighbor4_struct;
	neighbor_struct_t* neighbor6_struct;
	struct rte_acl_ctx *cur_acx_ipv4, *new_acx_ipv4;
	struct rte_acl_ctx *cur_acx_ipv6, *new_acx_ipv6;
	uint32_t kni_rate_limit_cur;

	// counter for each lower part of
	// dest ipv4 addrs in ratelimited cidr ranges
	uint32_t rlimit4_cur[MAX_RLIMIT_RANGE][MAX_RLIMIT_RANGE_HOST];
	uint32_t
	    rlimit6_cur[NEI_NUM_ENTRIES];  // counter for each ipv6 neighbor
} __rte_cache_aligned;

extern struct lcore_conf lcore_conf[RTE_MAX_LCORE];

struct rlimit6_data {
	struct in6_addr addr;
	uint32_t num;
};

union rlimit_addr {
	struct {
		uint32_t host : (32 - RLIMIT_RANGE);
		uint32_t network : RLIMIT_RANGE;
	};
	uint32_t addr;
};
extern uint8_t rlimit4_lookup_table[NB_SOCKETS][MAX_RLIMIT_RANGE_NET];
extern struct rlimit6_data rlimit6_lookup_table[NB_SOCKETS][NEI_NUM_ENTRIES];
extern uint32_t rlimit4_max[NB_SOCKETS][MAX_RLIMIT_RANGE]
			   [MAX_RLIMIT_RANGE_HOST];
extern uint32_t rlimit6_max[NB_SOCKETS][NEI_NUM_ENTRIES];

#define CMD_LINE_OPT_CONFIG "config"
#define CMD_LINE_OPT_KNICONFIG "kniconfig"
#define CMD_LINE_OPT_CALLBACK_SETUP "callback-setup"
#define CMD_LINE_OPT_UNIXSOCK "unixsock"
#define CMD_LINE_OPT_RULE_IPV4 "rule_ipv4"
#define CMD_LINE_OPT_RULE_IPV6 "rule_ipv6"
#define CMD_LINE_OPT_ACLAVX2 "aclavx2"
#define CMD_LINE_OPT_PROMISC "promiscuous"
#define CMD_LINE_OPT_NO_NUMA "no-numa"
#define CMD_LINE_OPT_ENABLE_JUMBO "enable-jumbo"
#define CMD_LINE_OPT_MAXPKT_LEN "max-pkt-len"
#define CMD_LINE_OPT_PORTMASK "portmask"
#define CMD_LINE_OPT_CONFIGFILE "configfile"
#define CMD_LINE_OPT_KNI_RATE_LIMIT "kni_rate_limit"
#define CMD_LINE_OPT_RATE_LIMIT "rate_limit"

#define FILE_MAIN_CONFIG "pktj"

#define MAX_LCORE_PARAMS 1024
#define MAX_JUMBO_PKT_LEN 9600

#endif
